c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,379 @@
+    }
+
+    fn record(&mut self, loc: Location, sub_loc: &[SubLoc], ex: &hir::Expr) {
+        self.record_desc(loc, sub_loc, ex, MirOriginDesc::Expr);
+    }
+
+    fn record_desc(
+        &mut self,
+        loc: Location,
+        sub_loc: &[SubLoc],
+        ex: &hir::Expr,
+        desc: MirOriginDesc,
+    ) {
+        let origin = MirOrigin {
+            hir_id: ex.hir_id,
+            span: ex.span,
+            desc,
+        };
+        match self.mir_map.entry((loc, sub_loc.to_owned())) {
+            Entry::Vacant(e) => {
+                e.insert(origin);
+            }
+            Entry::Occupied(e) => {
+                if *e.get() != origin {
+                    let old_origin = e.get().clone();
+                    error!(
+                        "conflicting origins for {:?} {:?} ({:?})\n\
+                            origin 1 = {:?}\n\
+                            origin 2 = {:?}\n\
+                            expr 2 = {:?}",
+                        loc,
+                        sub_loc,
+                        self.location_span(loc),
+                        origin,
+                        old_origin,
+                        ex,
+                    );
+                }
+            }
+        }
+    }
+
+    fn get_sole_assign(
+        &self,
+        locs: &[Location],
+    ) -> Option<(Location, mir::Place<'tcx>, &'a mir::Rvalue<'tcx>)> {
+        if locs.len() != 1 {
+            return None;
+        }
+        self.get_last_assign(locs)
+    }
+
+    fn get_last_assign(
+        &self,
+        locs: &[Location],
+    ) -> Option<(Location, mir::Place<'tcx>, &'a mir::Rvalue<'tcx>)> {
+        let &loc = locs.last()?;
+        let stmt = self.mir.stmt_at(loc).left()?;
+        match stmt.kind {
+            mir::StatementKind::Assign(ref x) => Some((loc, x.0, &x.1)),
+            _ => None,
+        }
+    }
+
+    fn get_last_call(
+        &self,
+        locs: &[Location],
+    ) -> Option<(
+        Location,
+        mir::Place<'tcx>,
+        &'a mir::Operand<'tcx>,
+        &'a [mir::Operand<'tcx>],
+    )> {
+        let &loc = locs.last()?;
+        let term = self.mir.stmt_at(loc).right()?;
+        match term.kind {
+            mir::TerminatorKind::Call {
+                ref func,
+                ref args,
+                destination,
+                ..
+            } => Some((loc, destination, func, args)),
+            _ => None,
+        }
+    }
+
+    fn should_ignore_statement(&self, loc: Location) -> bool {
+        if let Some(stmt) = self.mir.stmt_at(loc).left() {
+            match stmt.kind {
+                mir::StatementKind::FakeRead(..)
+                | mir::StatementKind::StorageLive(..)
+                | mir::StatementKind::StorageDead(..)
+                | mir::StatementKind::Nop => return true,
+                _ => {}
+            }
+        }
+        false
+    }
+
+    fn is_var(&self, pl: mir::Place<'tcx>) -> bool {
```

Comment:
Does this need a `&self` arg?

Reply:
I've removed it.  (In a previous version, this code used an `is_temp_var` method, which does need access to `self.mir`.)